### ISSUE TITLE
JXCore Plesk Extension : Feature Request (Multiple Package Installation)

### ISSUE THREAD

**Issue by gopumon** [jxcore/jxcore#602]
I have some feature request for Plesk extension's NPM section.

1. Multiple Package Installation (for example : coma seperated package names will be parsed and installed in one go. This will help developers while deploying their node application).
2. Check for package update and provision to update the package (A feature to notify user that the npm package have an update available and A button to update it)

**Comment by ktrzeciaknubisa** [comment]
Thanks.

The 1st would be fairy easy to do. Npm already can install multiple packages from a single command line. In fact it already works though JXcore extension, however displays misleading error. 
Plus, probably we just need to strip any extra switches e.g. `--global`, `--save` etc. that user might provide too. Is there any reason for allowing extra switches?

The 2d request is also doable, however I would skip the "notifying" part. Instead, a button "Check for updates" which would check and display available module's version for each row in table. Plus extra "Update" button per each row. How does it sound?

**Comment by obastemur** [comment]
We could run 'jx install' on website's root folder per request?

**Comment by ktrzeciaknubisa** [comment]
> We could run 'jx install' on website's root folder per request?

@obastemur is this a feature request? Nice one. Right now we're installing npm modules at `jxconfig.globalModulePath` and only admin can do that. Allowing the users to install things for themselves makes sense.

**Comment by obastemur** [comment]
No. This doesn't override any security feature.

If a particular user is able to install modules from panel, same panel should give an option to call `jx install` on root.

**Comment by ktrzeciaknubisa** [comment]
I'm sorry, but I don't follow :)

> If a particular user is able to install modules from panel

Particular user is **not** able. Only admin. And not per website.

**Comment by obastemur** [comment]
Imagine that admin needs to install 50 modules one by one ;)

Anyway, this can be nice to have but definitely not a priority ATM.

**Comment by gopumon** [comment]
As far as I can understand JXCore have npm modules in global scope. So admin will install npm package for users. Then in a scenario where two users need two different versions on same npm package then the last installed version will be available to both the users.

So moving the NPM inside domain is a good option.

**Comment by gopumon** [comment]
I just noticed... all JXCore features can be accessed by user with all subscription access (admin) and user with a single subscription.

An user with a single subscription can access and modify other subscriptions.

Admin should have all the access. 

But for user with a single subscription
- JXCore Configuration tab should not be accessible
- Domain tab should only list domains under their subscription
- Ideally the NPM should come under Domains (the npm packages will be installed for each domains)
- Subscription tab should not be accessible

**Comment by ktrzeciaknubisa** [comment]
I believe you're mistaken. Please double check if your single subscription user is does not have admin rights.

**Comment by gopumon** [comment]
@ktrzeciaknubisa I checked here. Its behaving like that. By the way I have plesk 12 with "Web Admin Edition" license. But I still have some subscriptions (Webspaces) and I tried with an user with a single subscription to modify other subscriptions in JXCore subscription tab.

The user is an Administrator who have access to a single subscription.

**Comment by ktrzeciaknubisa** [comment]
> Domain tab should only list domains under their subscription

By "single subscription" you mean a single value chosen (instead of "All") in a combo on user config page, e.g. like below?

![plesk-user](https://cloud.githubusercontent.com/assets/5946844/10515588/6c35ebbe-7354-11e5-8937-468aaf65981e.png)

**Comment by gopumon** [comment]
Yes! Thats exactly what I meant.

**Comment by ktrzeciaknubisa** [comment]
Ok. Thanks. I'm going to see that. 

However:

> * JXCore Configuration tab should not be accessible
> * Subscription tab should not be accessible

Access to both of them is determined by being an admin, not by accessibility to one or all subscriptions.

I believe that if you create Customer, Reseller or just subscription User, they cannot see those tabs.

**Comment by gopumon** [comment]
My mediatemple (shared) DV have a 'Plesk 12 Web Host Edition' license. Where I can manage Resellers and Customers. But When I add the JXCore extension it is not going beyond the "Terms and Conditions" page. Coming back to the same "Terms and Conditions" page when I hit OK.

My personal VPS only have 'Plesk 12 Web Admin Edition" license, so I wont be able to create Customers or Resellers.

**Comment by ktrzeciaknubisa** [comment]
>  When I add the JXCore extension it is not going beyond the "Terms and Conditions" page

You may open separate issue for that and post platform info.

> My personal VPS only have 'Plesk 12 Web Admin Edition" license, so I wont be able to create Customers or Resellers.

But probably you can create new users for the subscriptions?

**Comment by gopumon** [comment]
Yes I have user with access to single subscription (webspace).

But they can access other subscriptions in JXCore.

**Comment by ktrzeciaknubisa** [comment]
@gopumon I've resolved "Access to subscriptions" issue. You can test it from here: https://github.com/ktrzeciaknubisa/jxcore-issues/raw/master/plesk/dev/JXcore_Plesk0274.zip

**Comment by ktrzeciaknubisa** [comment]
The above solution will work only for Plesk 12+ ( @gopumon  that's your version).
For lower releases this still needs to be implemented. I'm waiting for Plesk's support advices in this matter: https://github.com/plesk/pm-docs/issues/1 .

**Comment by gopumon** [comment]
I installed the version and tested with an user with a single subscription. I found the following flaws.
- All subscriptions are still accessible through 'Subscriptions' tab 
- Domains Tab have only the main domain, subdomains listed under the same subscriptions are note listed in it.
- All enabled domains are accessible through 'Subscriptions' tab, in 'Domains' column.

**Comment by ktrzeciaknubisa** [comment]
Thank you. I forgot that your user is an admin, even with subscription access restriction. BTW. how are you defining him as admin? Is this a "User Role" combo?

**Comment by gopumon** [comment]
Yes! thats "User Role".

**Comment by ktrzeciaknubisa** [comment]
hi @gopumon . I've improved extension's behavior for users (even for admins) with subscription access restriction.

- [x] Works for Plesk 12+ but also below.
- [x] Subdomains are visible too.
- [x] The subscription tab is still visible only for admins.
- [x] If an admin has subscription access restriction - JXcore tab is not visible.

You may fetch the file from same link ([JXcore_Plesk0274.zip](https://github.com/ktrzeciaknubisa/jxcore-issues/raw/master/plesk/dev/JXcore_Plesk0274.zip)) end test it out.

**Comment by gopumon** [comment]
I tested, I can access everything with the restricted admin user through 'custom button' JXCore creates while installing.

**Comment by ktrzeciaknubisa** [comment]
Oh, I've found single wrong condition. Would you re-download and try again?

**Comment by gopumon** [comment]
Yes! This worked. 😃

The actual feature request was to have a 'Bulk install of npm package' and a 'check for update' for existing package. As it evolved, I think it is better to have npm packages for each domains.

The scenario made me think of this was: one of my application need an npm package with a specific version and other needs a greater version, if I update, then my first application will not work and if I don't I cannot run my second application.

Thank You.

**Comment by ktrzeciaknubisa** [comment]
Thank you, too. Yes, I understood the npm bulk install idea / install per domains. Those topics are related and will be the next thing to implement. In fact I already started to work on this.

**Comment by ktrzeciaknubisa** [comment]
Now this is a **huge** update: [JXcore_Plesk0275.zip](https://github.com/ktrzeciaknubisa/jxcore-issues/raw/master/plesk/dev/JXcore_Plesk0275.zip).

- [x] Multiple packages installation
- [x] Collapsible preview for *npm-debug.log* (can be seen once the `npm install error` happens, e.g. with non existing module/version)
- [x] New buttons above the modules table list: `Check for updates`, `Update All`
- [x] New column in table list showing version status or `Update to x.x.x` button
- [x] NPM modules installable per domains. For admin available always. For users needs to be enabled first by admin on `JXcore Configuration` page.

@gopumon Would you want to test it all? Let me now, if you'll find anything :)

**Comment by gopumon** [comment]
Tested everything with 100% success!! Awesome Job @ktrzeciaknubisa !! 👏

I liked that `npm-debug.log` preview, sure it is going to help developers a lot.

Initially the new column after remove shows *Check for updates* when I clicked the button `Check for updates` it shows some packages to update and other *This is the latest*; Will this change back to *Check for updates* after a period of time? or status will remain same and user need to hit `Check for updates` to know if there is any update? [*This is not important, just curious to know how this works*]

**Comment by ktrzeciaknubisa** [comment]
Thanks @gopumon !

> Will this change back to Check for updates after a period of time? 

Well, the *This is the latest* etc. status will be valid for 30 minutes. After that it will switch to null (*Check for updates*). I was tempted to implement background checking, but a) depending on how much modules you have in folder, this can be cpu taker, b) it should iterate all domain folders now, not just one *node_modules*, as we had before. So I left it for manual check.

**Comment by gopumon** [comment]
All good!! Manual check is fine.

**Comment by ktrzeciaknubisa** [comment]
BTW. whenever you install, update or remove any of npm modules, the *Check for updates* runs automatically. However I noticed that it was deleting *npm-debug.log* so I fixed this and posted the update under same link above. 

Also:

- [x] Added `Remove all` button for dispatching entire *node_modules*
- [x] Slightly improved the way *npm-debug.log* bar is displayed

I think that would be all. We're going to wait a bit in case if other users will report something. Then we will give it for the official release.

Thanks @gopumon !
